crates/iroha_core/src/sumeragi/main_loop.rs

```diff
@@ -882,9 +882,15 @@ impl Sumeragi {
         let view_change_in_progress = self.topology.view_change_index() > 0;
         let block_time = state.world.view().parameters.sumeragi.block_time();
         let deadline_reached = self.round_start_time.elapsed() > block_time;
+
         let tx_cache_non_empty = !self.transaction_cache.is_empty();
+        let prev_block_is_empty = state
+            .view()
+            .latest_block()
+            .map_or(true, |block| block.is_empty());
+        let block_expected = tx_cache_non_empty || !prev_block_is_empty;
 
-        if tx_cache_full || tx_cache_non_empty && (view_change_in_progress || deadline_reached) {
+        if tx_cache_full || block_expected && (view_change_in_progress || deadline_reached) {
```

Comment:
Nice, this is a correct grouping. If `tx_cache_non_empty` were not included in `block_expected`, the only way for a block following an empty block to proceed would be `tx_cache_full`